text/0000-app-mixins.md

```diff
@@ -0,0 +1,241 @@
+1. During the lifecycle's build phase, the stackpacks that passed detection will run against the build and run images accordingly (see details below). All stackpacks will be run before the user's buildpacks.
+
+## Stackpacks
+
+ A stackpacks is a special case of buildpack that has the following properties:
+
+* Is run as the `root` user
+* Configured with `privileged = true` in the `buildpack.toml`
+* Can only create one layer
+* Does not have access to the application source code
+* Is run before all regular buildpacks
+* Is run against both the build and run images
+* Is distributed with the builder image
+* May not write to the `/layers` directory
+
+The stackpack interface is identical to the buildpack interface (i.e. the same `bin/detect` and `bin/build` scripts are required). However, some of the context it is run in is different from regular buildpacks.
+
+For each stackpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the stackpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively, a platform may store a new stack image to cache the changes. All of the captured changes will be included in a single layer produced as output from the stackpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+A stackpack is included in a builder by defining it in the `builder.toml` in the `[[stack.buildpacks]]` array of tables:
+
+```
+[[stack.buildpacks]]
+id = "<stackpack id>"
+uri = "<uri to stackpack>"
+```
+
+Each stackpack included in the builder will execute until all of a project's mixins have been provided.
+
+The stackpack's snapshot layer may be modified by writing a `launch.toml` file. The `[[processes]]` and `[[slices]]` tables may be used as normal, and a new `[[excludes]]` table will be introduced with the following schema:
+
+```
+[[excludes]]
+paths = ["<sub-path glob>"]
+cache = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. Then, the rebase operation can be performed as normal, while including the stackpack layers as part of the stack. This will be made possible by including the stackpack in the run-image, but because the stackpack detect phase is not run, the operation does not need access to the application source.
+
+## Example: Apt Buildpack
+
+A buildpack that installs a custom certificate would have a `buildpack.toml` like this:
+
+ ```toml
+[buildpack]
+id = "example/apt"
+privileged = true
+mixins = ["[^=]+"]
```

Comment:
This `mixins` pattern seems quite broad - won't it mean this buildpack matches against too many mixin names? (It's quite possible I'm misunderstanding how the matching works here)

Reply:
The mixins concept is intentionally flexible as a whole. But it can be more narrowly scoped for each stack. For example, [a mixin is defined as a "package" for the bionic stack](https://github.com/buildpacks/rfcs/blob/main/text/0031-bionic-mixins.md). Because the stackpack in this example only works with the bionic stack, it can safely assume the mixin is a package.

Reply:
Addressed in 8cd657fc613769f29f84ca88673415504d57ea4c